firmware/targets/f7/furi-hal/furi-hal-spi.c

```diff
@@ -11,41 +11,55 @@
 
 #define TAG "FuriHalSpi"
 
+static void furi_hal_spi_bus_init(const FuriHalSpiBus* bus) {
+    furi_assert(bus);
+    furi_assert(bus->mutex);
+
+    if (*bus->mutex == NULL) {
+        *bus->mutex = osMutexNew(NULL);
+    }
+
+    hal_gpio_init_ex(bus->miso, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, bus->alt_fn);
+    hal_gpio_init_ex(bus->mosi, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, bus->alt_fn);
+    hal_gpio_init_ex(bus->clk, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, bus->alt_fn);
+}
+
+static void furi_hal_spi_device_cs_init(const FuriHalSpiDevice* device) {
+    furi_assert(device);
+
+    hal_gpio_init(
+        device->chip_select,
+        GpioModeOutputPushPull,
+        GpioPullNo,
+        GpioSpeedVeryHigh
+    );
+    hal_gpio_write(device->chip_select, true);
+}
+
 void furi_hal_spi_init() {
-    // Spi structure is const, but mutex is not
-    // Need some hell-ish casting to make it work
-    *(osMutexId_t*)spi_r.mutex = osMutexNew(NULL);
-    *(osMutexId_t*)spi_d.mutex = osMutexNew(NULL);
-    // 
     for (size_t i=0; i<FuriHalSpiDeviceIdMax; ++i) {
-        hal_gpio_write(furi_hal_spi_devices[i].chip_select, true);
-        hal_gpio_init(
-            furi_hal_spi_devices[i].chip_select,
-            GpioModeOutputPushPull,
-            GpioPullNo,
-            GpioSpeedVeryHigh
-        );
+        // for devices that are native to the bus, configure CS immediately
+        if (furi_hal_spi_devices[i].main_bus_config == NULL) {
+            furi_hal_spi_device_cs_init(&furi_hal_spi_devices[i]);
+        }
     }
 
-    hal_gpio_init_ex(&gpio_spi_r_miso, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, GpioAltFn5SPI1);
-    hal_gpio_init_ex(&gpio_spi_r_mosi, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, GpioAltFn5SPI1);
-    hal_gpio_init_ex(&gpio_spi_r_sck, GpioModeAltFunctionPushPull, GpioPullNo, GpioSpeedVeryHigh, GpioAltFn5SPI1);
-
-    hal_gpio_init_ex(&gpio_spi_d_miso, GpioModeAltFunctionPushPull, GpioPullUp, GpioSpeedVeryHigh, GpioAltFn5SPI2);
-    hal_gpio_init_ex(&gpio_spi_d_mosi, GpioModeAltFunctionPushPull, GpioPullUp, GpioSpeedVeryHigh, GpioAltFn5SPI2);
-    hal_gpio_init_ex(&gpio_spi_d_sck, GpioModeAltFunctionPushPull, GpioPullUp, GpioSpeedVeryHigh, GpioAltFn5SPI2);
+    furi_hal_spi_bus_init(&spi_r);
```

Comment:
Code to reconfigure gpio to analog on pin change is missing.
Configuring multiple pins to work with the same bus is going to cause issues.

I was planning to redesign SPI/I2C HAL and looks like the time has come.

I'm closing this PR, it will be easier to implement and test it internally before releasing.
Also if you have some specific use cases we can add them to tests.

Anyway good job.

Reply:
I pushed an update implementing switching to analog on bus re-configuration.
Any other issues with my implementation?

> Also if you have some specific use cases we can add them to tests.
We need a convenient way to access external SPI, using custom `LL_SPI_InitTypeDef` configuration.
My PR fully implements an API for that.

Reply:
> Any other issues with my implementation?

Not sure yet, need some time for research.

I was thinking to move this part and its corresponding routines:

```
     .miso=&gpio_spi_r_miso,
     .mosi=&gpio_spi_r_mosi,
     .clk=&gpio_spi_r_sck,
     .alt_fn=GpioAltFn5SPI1,
```

to separate structure or callback. Also it will be nice to gather all use cases first. Yours is quite standard.

Do you have flipper? Do you want to continue working on this PR?

Reply:
>I was thinking to move this part and its corresponding routines:

Well, almost all pin configurations for SPI are static, only exception being PA4/PB2 on external connector that are both capable of running in SPI1_NSS mode. But, to reduce complexity and stay in line with current code, I made a fixed definition for external bus. 

> Do you have flipper? Do you want to continue working on this PR?

No, I don't have a real device (still waiting for my kickstarted bundle to arrive) -- and this PR was a part of preparation for our team's task https://github.com/flipperdevices/flipper-hackathon-moscow/issues/3. 
I'd love to contribute to the repo on this subject and wish I could test my changes myself, not bothering @aprosvetova to run my builds.

Also, for our team's task, we'd need a way of accessing SPI3 -- current HAL implementation only configures a single internal bus.

Reply:
I've scheduled this changes for current sprint, they'll arrive to `dev` till next Wednesday.

Also you visit our office and test this feature with real hardware.